zebra-state/src/service/read/difficulty.rs

```diff
@@ -308,16 +319,28 @@ fn adjust_difficulty_and_time_for_testnet(
         .expect("a valid block time minus a small constant is in-range");
 
     if result.cur_time <= local_std_difficulty_limit {
-        // Standard difficulty: the max time needs to exclude min difficulty blocks
-        result.max_time = std_difficulty_max_time;
+        // Standard difficulty: the cur and max time need to exclude min difficulty blocks
+
+        // The maximum time can only be decreased, and only as far as min_time.
+        // The old minimum is still required by other consensus rules.
+        result.max_time = min(std_difficulty_max_time, result.max_time);
+        result.max_time = max(result.min_time, result.max_time);
+
+        // The current time only needs to be decreased if the max_time decreased past it.
+        // Decreasing the current time can't change the difficulty.
+        result.cur_time = result.cur_time.clamp(result.min_time, result.max_time);
     } else {
-        // Minimum difficulty: the min and cur time need to exclude min difficulty blocks
-        result.min_time = min_difficulty_min_time;
-        if result.cur_time < min_difficulty_min_time {
-            result.cur_time = min_difficulty_min_time;
-        }
+        // Minimum difficulty: the min and cur time need to exclude std difficulty blocks
+
+        // The minimum time can only be increased, and only as far as max_time.
+        // The old maximum is still required by other consensus rules.
+        result.min_time = max(min_difficulty_min_time, result.min_time);
+        result.min_time = min(result.max_time, result.min_time);
+
+        // The current time only needs to be increased if the min_time increased past it.
+        result.cur_time = result.cur_time.clamp(result.min_time, result.max_time);
 
-        // And then the difficulty needs to be updated for cur_time
+        // And then the difficulty needs to be updated for cur_time.
         result.expected_difficulty = AdjustedDifficulty::new_from_header_time(
             result.cur_time.into(),
```

Comment:
No change needed here but I wonder if we could set the `min_time` to `min_difficulty_min_time` as long as `max_time` is ahead of `std_difficulty_max_time` then use `min_difficulty_min_time` here instead of `cur_time`.

i.e. tell miners to use the minimum difficulty threshold then submit blocks with a timestamp ahead of their local clock.

https://github.com/ZcashFoundation/zebra/blob/691738226938a66c45527c5004b5710a294c6f0d/zebra-state/src/service/check.rs#L101-L106

Reply:
> I wonder if we could set the `min_time` to `min_difficulty_min_time` as long as `max_time` is ahead of `std_difficulty_max_time`

You might have already worked this out, but this is almost always true, at least for the first few blocks, or if a node is a minority miner. The `max_time` is 90 minutes ahead of the median, and the `std_difficulty_max_time` is 7.5 minutes ahead of the previous block.

So if we want to test in both modes, we need to support mining both standard and minimum difficulty blocks.